### PR TITLE
Added conditional for cloud formation 

### PR DESCRIPTION
**Pull request by yindia**
- Improved security by adding conditional IAM permission restriction based on `ManagedByUnion` tag in cloud formation.
- Fixed documentation by updating the correct S3 bucket URL for CloudFormation one-click button.
- Renamed roles and policy names for better clarity and consistency in the project. 
  - `unionai-provisioner`
  - `unionai-updater`
  - `unionai-reader`

### CODE REVIEW

File: union-ai-admin/aws/script/generate.py [hunk 1]
```diff
@@ -183,39 +185,90 @@ def create_read_policy(role_type):
...
-                        Sub("arn:aws:iam::${AWS::AccountId}:role/*"),
                         Sub("arn:aws:iam::${AWS::AccountId}:instance-profile/*"),
                     ],
                 ),
+                Statement(
+                    Effect=Allow,
+                    Action=[
+                        Action("ec2", "DescribeVpcAttribute"),
+                        Action("ec2", "AssociateAddress"),
+                        Action("ec2", "DisassociateAddress"),
```
Comment (bdconnolly) [inline union-ai-admin/aws/script/generate.py:262]: Is this correct? Should the reader policy have permissions to AssociateAddress and DisassociateAddress?

File: union-ai-admin/aws/script/generate.py [hunk 2]
```diff
@@ -183,39 +185,90 @@ def create_read_policy(role_type):
...
+                        Sub("arn:aws:iam::${AWS::AccountId}:role/*userflyterole*"),
+                        Sub("arn:aws:iam::${AWS::AccountId}:role/*adminflyterole*"),
+                        Sub("arn:aws:iam::${AWS::AccountId}:role/*fluentbitrole*"),
+                    ],
+                ),
+                Statement(
+                    Effect=Allow,
+                    Action=[
+                        Action("iam", "ListOpenIDConnectProviderTags"),
+                        Action("iam", "ListInstanceProfileTags"),
```
Comment (bdconnolly) [inline union-ai-admin/aws/script/generate.py:250]: Can these be merged with the statement above which has 

```
Action("iam", "GetOpenIDConnectProvider"),
Action("iam", "GetInstanceProfile"),
```
They look to have the same resource specified.

File: union-ai-admin/aws/script/generate.py [hunk 3]
```diff
@@ -563,84 +673,60 @@ def create_provisioner_policy(role_type):
...
-                            "arn:aws:eks:${AWS::Region}:${AWS::AccountId}:nodegroup/opta-*/opta-*/*"
-                        ),
-                        Sub(
-                            "arn:aws:eks:${AWS::Region}:${AWS::AccountId}:nodegroup/opta-*"
-                        ),
-                    ],
-                ),
                 Statement(
                     Effect=Allow,
                     Action=[
```
Comment (bdconnolly) [inline union-ai-admin/aws/script/generate.py:732]: We should not use `Action("s3", "*")`. We should be able to define which s3: actions we need to perform.

File: union-ai-admin/aws/script/generate.py [hunk 4]
```diff
@@ -655,22 +741,19 @@ def create_admin_policy(role_type):
...
                 ),
                 Statement(
                     Effect=Allow,
                     Action=[
                         Action("ec2", "EnableEbsEncryptionByDefault"),
+                        Action("ec2", "ModifyEbsDefaultKmsKeyId"),
+                        Action("ec2", "DisableEbsEncryptionByDefault"),
                         Action("ec2", "GetEbsEncryptionByDefault"),
                     ],
                     Resource=["*"],
```
Comment (bdconnolly) [inline union-ai-admin/aws/script/generate.py:759]: We can set a condition on this for the region where we will be enabling this. See https://docs.aws.amazon.com/service-authorization/latest/reference/list_amazonec2.html

**Review by bdconnolly — COMMENTED** [review]
It does not look the unionai-reader role has permissions to view the read  or list an objects on the S3 buckets. Is that by design?

**Comment by bdconnolly** [comment]
In the terraform policy, we should not use Action("s3", "*"). We should define which s3: actions we need permissions to perform.

File: union-ai-admin/aws/script/generate.py (hunk 1 above)
Comment (bdconnolly) [reply union-ai-admin/aws/script/generate.py:262]: This section of code comes the from the function defined with `def create_read_policy(role_type):`  This policy is attached the unionai-reader role and the other roles.

File: union-ai-admin/aws/script/generate.py [hunk 5]
```diff
@@ -183,39 +203,74 @@ def create_read_policy(role_type):
...
                         Sub("arn:aws:iam::${AWS::AccountId}:instance-profile/*"),
                     ],
                 ),
+                Statement(
+                    Effect=Allow,
+                    Action=[
+                        Action("ec2", "DescribeVpcAttribute"),
+                    ],
+                    Resource=["*"],
+                    Condition=Condition(
```
Comment (ctso) [inline union-ai-admin/aws/script/generate.py:267]: No need to do this now, but we could probably move this `Condition` into a constant so we can re-use it and repeat ourselves less in this codebase.

**Review by ctso — APPROVED** [review]

**Pull request merged**